display_list/display_list_canvas_unittests.cc

```diff
@@ -1121,6 +1121,67 @@ class CanvasCompareTester {
       }
     }
 
+    {
+      // Being able to see a dilate requires some non-default attributes,
+      // like a non-trivial stroke width and a shader rather than a color
+      // (for drawPaint) so we create a new environment for these tests.
+      RenderEnvironment dilate_env = RenderEnvironment::MakeN32();
+      CvSetup cv_dilate_setup = [=](SkCanvas*, SkPaint& p) {
+        p.setShader(testImageColorSource.skia_object());
+        p.setStrokeWidth(5.0);
+      };
+      DlRenderer dl_dilate_setup = [=](DisplayListBuilder& b) {
+        b.setColorSource(&testImageColorSource);
+        b.setStrokeWidth(5.0);
+      };
+      dilate_env.init_ref(cv_dilate_setup, testP.cv_renderer(),
+                          dl_dilate_setup);
+      DlDilateImageFilter filter_5(5.0, 5.0);
+      BoundsTolerance dilate5Tolerance = tolerance.addBoundsPadding(21, 21);
+      RenderWith(testP, dilate_env, dilate5Tolerance,
+                 CaseParameters(
+                     "ImageFilter == Dilate 5",
+                     [=](SkCanvas* cv, SkPaint& p) {
+                       cv_dilate_setup(cv, p);
+                       p.setImageFilter(filter_5.skia_object());
+                     },
+                     [=](DisplayListBuilder& b) {
+                       dl_dilate_setup(b);
+                       b.setImageFilter(&filter_5);
+                     }));
+    }
+
+    {
+      // Being able to see an erode requires some non-default attributes,
+      // like a non-trivial stroke width and a shader rather than a color
+      // (for drawPaint) so we create a new environment for these tests.
+      RenderEnvironment erode_env = RenderEnvironment::MakeN32();
+      CvSetup cv_erode_setup = [=](SkCanvas*, SkPaint& p) {
+        p.setShader(testImageColorSource.skia_object());
+        p.setStrokeWidth(6.0);
+      };
+      DlRenderer dl_erode_setup = [=](DisplayListBuilder& b) {
+        b.setColorSource(&testImageColorSource);
+        b.setStrokeWidth(6.0);
+      };
+      erode_env.init_ref(cv_erode_setup, testP.cv_renderer(), dl_erode_setup);
+      // do not erode too much, because some tests assert there are enough
+      // pixels that are changed.
+      DlErodeImageFilter filter_1(1.0, 1.0);
+      BoundsTolerance erode5Tolerance = tolerance.addBoundsPadding(21, 21);
```

Comment:
Thank you for updating these tests, BTW, many of the added capabilities here don't really add tests to this file. It's not a hard requirement, but one thing it does is to figure out if we're computing accurate bounds.

To that end, the tolerance object is primarily meant to indicate "the bounds you compute may be conservative and it is too hard to come up with a test case that demonstrates the pessimal bounds case that drives the conservative computations.

In this case, was 21 padding really needed? The bounds would be enlarged by only 1 pixel on each side and the bounds computation on this type of filter should be really accurate. I'm surprised any bounds padding is needed at all. Was this perhaps from the time when DlDilateFilter was computing bounds with an extra `* 3`?